Integration with App Switch Eligibility, Launching app switch

kgangineni opened the pull request:
### Summary of changes

1. Introduces Boolean `appSwitchWhenEligible`, allowing the merchants to opt-in app switch flow
2. Makes patchCCOWithAppSwitchEligibility call to verify eligibility
3. Introduces callback for Vault and Checkout flows to enable callback flow for app switch

 ### Checklist


 - [ ] Added a changelog entry

### Authors
> List GitHub usernames for everyone who contributed to this pull request.

-

kgangineni committed:
⏺ Add OAuth2 client token fetching with suspend PayPalWebCheckoutClient

  - Implement FetchClientToken class for OAuth2 client credentials flow
  - Add headers parameter to APIRequest for Basic auth support
  - Convert PayPalWebCheckoutClient.start() to suspend function
  - Integrate token fetching before launching web checkout
  - Add comprehensive error handling for token fetch failures

kgangineni committed:
⏺ Implement app switch eligibility flow for PayPal checkout

  - GraphQL integration for PatchCcoWithAppSwitchEligibility
  - Integrate app switch eligibility, get redirect URL for launching checkout
  - Launch with redirect URL when available, fallback to web checkout
  - Create PayPalWebCheckoutCallback fun interface following CardApproveOrderCallback pattern
  - Add overloaded start() function accepting PayPalWebCheckoutCallback

kgangineni committed:
Refactor code to re-use constants in headers

kgangineni committed:
Merge remote-tracking branch 'origin/feature/app_switch' into BooleanToOptInAppSwitch

kgangineni committed:
- Improve null check in parsing response
- Move parsing response logic to PatchCCOWithAppSwitchEligibilityResponse class
- Move reading query and creating json object for request to PatchCcoWithAppSwitchEligibilityRequest class
- Improve test cases
- add paypalNativeAppInstalled param to graphQL request

kgangineni committed:
⏺ Implement app switch logic for vault operations and fix test issues

  - Add app switch eligibility check for vault functionality
  - Create PayPalWebVaultCallback interface for vault operation callbacks
  - Extend PayPalWebLauncher with launchWithUrl() method supporting TokenType mapping

kgangineni committed:
- Fix unit test for createLowScopedAccessToken()
- Add Change log entry for public API changes
- made items internal where possible

sshropshire commented on CorePayments/src/main/java/com/paypal/android/corepayments/api/PatchCCOWithAppSwitchEligibility.kt:
```diff
@@ -0,0 +1,76 @@
+            val token =
+                when (val tokenResult = authenticationSecureTokenServiceAPI.createLowScopedAccessToken()) {
+                    is APIResult.Success -> tokenResult.data
+                    is APIResult.Failure -> return APIResult.Failure(tokenResult.error)
```
🧶 : Is there a way to clean up this return statement? I understand what it's doing, but it seems sort of like an escape hatch. It might read a lot cleaner if we make the output of `when` be the return value of the `runCatching` block, rather than assigning the result of `when(){}` to `token`.

sshropshire commented on CorePayments/src/main/java/com/paypal/android/corepayments/model/PatchCcoWithAppSwitchEligibilityRequest.kt:
```diff
@@ -0,0 +1,69 @@
+        return JSONObject()
+            .put("query", query)
+            .put("variables", variablesJson)
+    }
```
I like this pattern. What if we move `create()` into a `companion object`? It doesn't seem to rely on instance state. Then `PatchCcoWithAppSwitchEligibilityRequest` can be a plain data object.

kgangineni replied:
That was the initial implementation, but variables made sense to be in the data class, that's the instance state used

sshropshire commented on CorePayments/src/main/java/com/paypal/android/corepayments/model/PatchCcoWithAppSwitchEligibilityResponse.kt:
```diff
@@ -0,0 +1,58 @@
+import com.paypal.android.corepayments.APIClientError
+import org.json.JSONObject
+
+data class PatchCcoWithAppSwitchEligibilityResponse(
```
q: Can we make these data objects be `internal`? Merchants shouldn't need to create these types.

kgangineni replied:
response is used in `PaypalWebCheckoutClient` but made it restricted to Library group

sshropshire commented on CorePayments/src/main/java/com/paypal/android/corepayments/api/PatchCCOWithAppSwitchEligibility.kt:
```diff
@@ -0,0 +1,76 @@
+        graphQLClient = GraphQLClient(coreConfig)
+    )
+
+    suspend operator fun invoke(
```
🧶: Should we use the [UseCase suffix naming convention](https://developer.android.com/topic/architecture/domain-layer#conventions) for this class?

kgangineni replied:
I'm kind of in between, if in demo app definitely useCase
CoPilot response: 
Use PatchCCOWithAppSwitchEligibility for a straightforward implementation.
Use PatchCCOWithAppSwitchEligibilityUseCase if it is in layered clean architecture

sshropshire replied:
Which do you think is best going forward?

kgangineni replied:
I think better to go with PatchCCOWithAppSwitchEligibility pattern, inside sdk, we can call it like a function

sshropshire commented on CorePayments/src/test/java/com/paypal/android/corepayments/api/PatchCCOWithAppSwitchEligibilityUnitTest.kt:
```diff
@@ -0,0 +1,302 @@
+            tokenType = TokenType.ORDER_ID,
+            merchantOptInForAppSwitch = true,
+            paypalNativeAppInstalled = true
+        )
```
```suggestion
        sut(
            context = context,
            orderId = testOrderId,
            tokenType = TokenType.ORDER_ID,
            merchantOptInForAppSwitch = true,
            paypalNativeAppInstalled = true
        )
```
🧶 : We shouldn't need to explicitly call `invoke` here, we can just use the `()` invocation operator.

sshropshire commented on PayPalWebPayments/src/main/java/com/paypal/android/paypalwebpayments/PayPalWebCheckoutClient.kt:
```diff
@@ -110,14 +128,43 @@ class PayPalWebCheckoutClient internal constructor(
      * @param request [PayPalWebVaultRequest] for vaulting PayPal as a payment method
      */
-    fun vault(
+    suspend fun vault(
```
We may need a `CHANGELOG` entry for this breaking change.

sshropshire commented on PayPalWebPayments/src/main/java/com/paypal/android/paypalwebpayments/PayPalWebCheckoutClient.kt:
```diff
@@ -110,14 +128,43 @@ class PayPalWebCheckoutClient internal constructor(
+            // if RedirectUrl not available fallback to web vault
+            if (!launchUrl.isNullOrEmpty()) {
+                payPalWebLauncher.launchWithUrl(
+                    activity,
```
Oh wait my bad, I forgot that BrowserSwitch requires an activity reference. Please disregard.

sshropshire commented on CHANGELOG.md:
````diff
@@ -1,5 +1,23 @@
 # PayPal Android SDK Release Notes
 
+```kotlin 
````
This should probably be in a migration guide rather than the CHANGELOG. We don't typically place code snippets in our release notes.

KunJeongPark commented on CorePayments/src/main/java/com/paypal/android/corepayments/model/PatchCcoWithAppSwitchEligibilityRequest.kt:
```diff
@@ -0,0 +1,69 @@
+
+    companion object {
+        const val INTEGRATION_ARTIFACT = "NATIVE_SDK" // todo: use Mobile SDK artifact after backend changes
+        const val INTEGRATION_CHANNEL = "PPCP_NATIVE_SDK"
```
would this translate to CCO field "channel" or is this something else?
This reminds me, we should request this right away to XOOS, to accommodate our integration artifact value

If so, I think our plan was to use "MOBILE_APP"

kgangineni replied:
`MOBILE_APP`? in Jira it's mentioned as MOBILE_SDK, guess whatever changes u made for updateConfig, we will go with same

KunJeongPark replied:
Yes, those values were changed in meeting with Nirvan and risk and compliance.
does that INTEGRATION_CHANNEL map to channel CCO field?
I didn't know that there was PPCP_NATIVE_SDK
Did you try running this with MOBILE_SDK and see if the graphQL fails or not?

kgangineni replied:
there is `PPCP_NATIVE_SDK` in graphql schema, `MOBILE_SDK` is not there, Yes I tried running, it gives graphql validation error
I will raise PR to add it to schema

KunJeongPark replied:
I am curious about the mapping of the fields here
I know integration artifact, that we need to raise the issue for this and you've already done that.
but not sure what` integration channel` is
that is different from xobuyernodeserv's UpdateClientConfig.
Is there an equivalent to channel for this endpoint?
Yes, I see that it belongs to the experimentationContext, separate from CCO values. 

I know that` xorouternodeweb` is looking for correct integration artifact value (MOBILE_SDK)
to determine whether to respect our CCO values. So I think that maybe related to
why we are not seeing these values persisted in session explorer for this flow.

sshropshire commented on CorePayments/src/main/java/com/paypal/android/corepayments/api/PatchCCOWithAppSwitchEligibility.kt:
```diff
@@ -0,0 +1,76 @@
+            val token =
+                when (val tokenResult = authenticationSecureTokenServiceAPI.createLowScopedAccessToken()) {
+                    is APIResult.Success -> tokenResult.data
+                    is APIResult.Failure -> throw tokenResult.error
```
qq: why do we need `runCatching` here? It doesn't seem like we're using any Java APIs.

kgangineni replied:
we're not using JAVA APIs, this is only to make sure to send result as `APIResult.Failure` rather than throwing exception

sshropshire replied:
Yeah that makes sense. If we aren't using any methods that `throw` can we remove `runCatching`?

kgangineni replied:
not just throwing error programatically, more of safe catching for unknown errors that we didn't thought of, kind of last catch block as fallback

sshropshire replied:
I do think we should return an `APIResult` here. It simplifies the API for the caller, they will only need a single `when` statement instead of a combination `when` + `try/catch`.

kgangineni replied:
removed `runCatching` & some of bloat in parsing logic, made it private function

sshropshire commented on CorePayments/src/main/java/com/paypal/android/corepayments/common/Headers.kt:
```diff
@@ -0,0 +1,7 @@
+package com.paypal.android.corepayments.common
+
+object Headers {
```
```suggestion
internal object Headers {
```

sshropshire commented on CorePayments/src/main/java/com/paypal/android/corepayments/api/PatchCCOWithAppSwitchEligibility.kt:
```diff
@@ -0,0 +1,76 @@
+import com.paypal.android.corepayments.model.TokenType
+import com.paypal.android.corepayments.model.Variables
+
+class PatchCCOWithAppSwitchEligibility internal constructor(
```
```suggestion
@RestrictTo(RestrictTo.Scope.LIBRARY_GROUP)
class PatchCCOWithAppSwitchEligibility internal constructor(
```

sshropshire commented on CorePayments/src/main/java/com/paypal/android/corepayments/model/TokenType.kt:
```diff
@@ -0,0 +1,8 @@
+package com.paypal.android.corepayments.model
+
+enum class TokenType {
```
```suggestion
@RestrictTo(RestrictTo.Scope.LIBRARY_GROUP)
enum class TokenType {
```

sshropshire commented on PayPalWebPayments/src/main/java/com/paypal/android/paypalwebpayments/PayPalWebCheckoutClient.kt:
```diff
@@ -58,31 +58,49 @@ class PayPalWebCheckoutClient internal constructor(
+            // Get RedirectUrl
+            val launchUrl = when (patchCcoResult) {
+                is APIResult.Success -> patchCcoResult.data.launchUrl
+                is APIResult.Failure -> null
```
```suggestion
                is APIResult.Failure -> buildFallbackURL()
```
🧶 : We could build the fallback URL here to make the implementation cleaner.

kgangineni replied:
ended cleaning up more, moved logic of buildUrl to `PayPalWebCheckoutClient` from `PayPalWebLauncher`
let me know what u think

kgangineni committed:
- refactored to create web checkout url in PayPalWebCheckoutClient to increase re-usability

sshropshire commented on CHANGELOG.md:
```diff
@@ -1,5 +1,12 @@
 # PayPal Android SDK Release Notes
 
+* added overloaded functions with callbacks for `start` and `vault` methods in
```
```suggestion
## unreleased
* added overloaded functions with callbacks for `start` and `vault` methods in
```

kgangineni committed:
- update CHANGELOG.md

sshropshire commented on PayPalWebPayments/src/main/java/com/paypal/android/paypalwebpayments/PayPalWebCheckoutClient.kt:
```diff
@@ -190,4 +229,69 @@ class PayPalWebCheckoutClient internal constructor(
+        }
+    }
+
+    private fun launchWithCallback(block: suspend () -> Unit) {
```
Does the block need to suspend here? I'd like to try removing it to simplify the code.

kgangineni replied:
yes, purpose is to take suspend function and make it callback so it can be called with java

sshropshire replied:
Do we need this method then? It seems like it does the exact same thing as
```kotlin
CoroutineScope(mainDispatcher).launch {
}
```
There's a slight readability gain for `launchWithCallback`, but when I see `CoroutineScope.launch()` I know exactly what's happening.

kgangineni replied:
I guess we can remove it for now, it's only used in this file

kgangineni committed:
remove runCatching, remove bloat in parsing of response

sshropshire commented on CHANGELOG.md:
```diff
@@ -1,5 +1,13 @@
+* modified `PayPalWebCheckoutRequest` to include `appSwitchWhenEligible` property to control app
+  switch behavior
+* Breaking Changes
+  * made start and vault functions in `PayPalWebCheckoutClient` suspend functions
```
```suggestion
* Add overloaded functions with callbacks for `start` and `vault` methods in
  `PayPalWebCheckoutClient`
* Modify `PayPalWebCheckoutRequest` to include `appSwitchWhenEligible` property to control app
  switch behavior
* Breaking Changes
  * Make start and vault functions in `PayPalWebCheckoutClient` suspend functions
```
Placing these in the [imperative mood](https://www.theserverside.com/blog/Coffee-Talk-Java-News-Stories-and-Opinions/git-commit-message-imperative-mood-past-tense-convention-standards-best-practice-subject) to align with git commit message style.

sshropshire commented on CorePayments/src/main/java/com/paypal/android/corepayments/graphql/GraphQLClient.kt:
```diff
@@ -36,13 +32,18 @@ class GraphQLClient internal constructor(
         val body = graphQLRequestBody.toString()
         val urlString = if (queryName != null) "$graphQLURL?$queryName" else graphQLURL
         val httpRequest = HttpRequest(URL(urlString), HttpMethod.POST, body, httpRequestHeaders)
+        headers?.forEach { (key, value) -> httpRequestHeaders.put(key, value) }
```
```suggestion
        headers?.forEach { (key, value) -> httpRequestHeaders.put(key, value) }
        val httpRequest = HttpRequest(URL(urlString), HttpMethod.POST, body, httpRequestHeaders)
```
I feel like it makes sense to add headers before constructing the `HttpRequest`. This will prevent bugs in the future if the `HttpRequest` constructor one day makes its own copy of `httpRequestHeaders`.

sshropshire commented on CorePayments/src/main/java/com/paypal/android/corepayments/model/PatchCcoWithAppSwitchEligibilityRequest.kt:
```diff
@@ -0,0 +1,69 @@
+            val query = when (val result = resourceLoader.loadRawResource(context, resId)) {
+                is LoadRawResourceResult.Success -> result.value
+
+                is LoadRawResourceResult.Failure -> throw APIClientError.graphQLRequestLoadError()
```
Can we replace `throw` here with a result type?

kgangineni replied:
returning nullable instead of throwing error

sshropshire commented on CorePayments/src/main/java/com/paypal/android/corepayments/model/PatchCcoWithAppSwitchEligibilityRequest.kt:
```diff
@@ -0,0 +1,69 @@
+    }
+
+    companion object {
+        const val INTEGRATION_ARTIFACT = "NATIVE_SDK" // todo: use Mobile SDK artifact after backend changes
```
```suggestion
        const val INTEGRATION_ARTIFACT = "NATIVE_SDK" // TODO: use Mobile SDK artifact after backend changes
```

sshropshire commented on CorePayments/src/test/java/com/paypal/android/corepayments/api/AuthenticationSecureTokenServiceAPIUnitTest.kt:
```diff
@@ -62,8 +63,8 @@ class AuthenticationSecureTokenServiceAPIUnitTest {
         assertEquals("application/x-www-form-urlencoded", headers["Content-Type"])
         assert(headers["Authorization"]!!.startsWith("Basic "))
```
```suggestion
        assertTrue(headers["Authorization"]!!.startsWith("Basic "))
```
I noticed this is Java `assert` rather than a JUnit `assertTrue`.

sshropshire commented on CorePayments/src/test/java/com/paypal/android/corepayments/api/AuthenticationSecureTokenServiceAPIUnitTest.kt:
```diff
@@ -62,8 +63,8 @@ class AuthenticationSecureTokenServiceAPIUnitTest {
 
-            assertTrue(result is CreateLowScopedAccessTokenResult.Success)
-            assertEquals("test-token", (result as CreateLowScopedAccessTokenResult.Success).token)
+            assertTrue(result is APIResult.Success)
```
We may need to auto format this file. The indentation seems to be off a bit on some lines.

sshropshire commented on CorePayments/src/test/java/com/paypal/android/corepayments/api/AuthenticationSecureTokenServiceAPIUnitTest.kt:
```diff
@@ -97,7 +98,7 @@ class AuthenticationSecureTokenServiceAPIUnitTest {
 
         // Then
-        assertTrue(result is CreateLowScopedAccessTokenResult.Failure)
+        assertTrue(result is APIResult.Failure)
```
Should we also assert the contents of the error message?

sshropshire commented on CorePayments/src/test/java/com/paypal/android/corepayments/api/AuthenticationSecureTokenServiceAPIUnitTest.kt:
```diff
@@ -220,11 +221,11 @@ class AuthenticationSecureTokenServiceAPIUnitTest {
         val authHeader = requestSlot.captured.headers!!["Authorization"]!!
         assert(authHeader.startsWith("Basic "))
```
```suggestion
        assertTrue(authHeader.startsWith("Basic "))
```

kgangineni committed:
* Minor improvements
  - Replace throw with null return in PatchCcoWithAppSwitchEligibilityRequest.create()
  - Add proper null handling for failed GraphQL request creation
  - Update unit tests to verify error codes and messages

sshropshire commented on CorePayments/src/main/java/com/paypal/android/corepayments/api/PatchCCOWithAppSwitchEligibility.kt:
```diff
@@ -0,0 +1,108 @@
+    ): APIResult<AppSwitchEligibility> {
+        val tokenResult = authenticationSecureTokenServiceAPI.createLowScopedAccessToken()
+        if (tokenResult is APIResult.Failure) {
+            return APIResult.Failure(tokenResult.error)
```
```suggestion
            return tokenResult
```
nit: We could technically forward the result in this case.

sshropshire commented on CorePayments/src/main/java/com/paypal/android/corepayments/model/PatchCcoWithAppSwitchEligibilityRequest.kt:
```diff
@@ -0,0 +1,68 @@
+
+        return JSONObject()
+            .put("experimentationContext", experimentationContext)
+            .put(
+                "integrationArtifact",
+                INTEGRATION_ARTIFACT
+            )
```
```suggestion
            .put("integrationArtifact", INTEGRATION_ARTIFACT)
```
nit: to help with readability.

sshropshire reviewed (approved):
LGTM

KunJeongPark commented on CorePayments/src/main/res/raw/graphql_query_patch_cco_app_switch_eligibility.graphql:
```diff
@@ -0,0 +1,34 @@
+      patchCcoInput: {
+        token: $token
+        clientConfig: {
+          integrationArtifact: $integrationArtifact
```
I think here is where you would also set channel: MOBILE_APP, userExperienceFlow: INCONTEXT, productFlow: HERMES.
I think product flow, that's the defaulted value in CSPS, but I just made it explicit in iOS just to be safe

KunJeongPark commented on CorePayments/src/main/res/raw/graphql_query_patch_cco_app_switch_eligibility.graphql:
```diff
@@ -0,0 +1,34 @@
+      appSwitchEligibilityInput: {
+        contextId: $contextId
+        experimentationContext: $experimentationContext
+        merchantOptInForAppSwitch: $merchantOptInForAppSwitch
```
I don't see any reference for experimentationContext or merchantOptInForAppSwitch
Are there docs for these?
